src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs

```diff
@@ -0,0 +1,232 @@
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.LanguageServer;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Newtonsoft.Json.Linq;
+using StreamJsonRpc;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.LanguageService
+{
+    internal class InProcLanguageServer
+    {
+        private readonly IDiagnosticService _diagnosticService;
+        private readonly JsonRpc _jsonRpc;
+        private readonly IThreadingContext _threadingContext;
+        private readonly LanguageServerProtocol _protocol;
+        private readonly Workspace _workspace;
+
+        private VSClientCapabilities? _clientCapabilities;
+
+        public InProcLanguageServer(Stream inputStream, Stream outputStream, LanguageServerProtocol protocol, Workspace workspace, IDiagnosticService diagnosticService, IThreadingContext threadingContext)
+        {
+            this._protocol = protocol;
+            this._workspace = workspace;
+            this._threadingContext = threadingContext;
+
+            this._jsonRpc = new JsonRpc(outputStream, inputStream, this);
+            this._jsonRpc.StartListening();
+
+            this._diagnosticService = diagnosticService;
+            this._diagnosticService.DiagnosticsUpdated += DiagnosticService_DiagnosticsUpdated;
+        }
+
+        /// <summary>
+        /// Handle the LSP initialize request by storing the client capabilities
+        /// and responding with the server capabilities.
+        /// The specification assures that the initialize request is sent only once.
+        /// </summary>
+        [JsonRpcMethod(Methods.InitializeName)]
+        public InitializeResult Initialize(JToken input)
+        {
+            // InitializeParams only references ClientCapabilities, but the VS LSP client
+            // sends additional VS specific capabilities, so directly deserialize them into the VSClientCapabilities
+            // to avoid losing them.
+            this._clientCapabilities = input["capabilities"].ToObject<VSClientCapabilities>();
+
+            return new InitializeResult
+            {
+                Capabilities = new VSServerCapabilities
+                {
+                    DocumentHighlightProvider = true,
+                    DocumentSymbolProvider = true,
+                    DocumentFormattingProvider = true,
+                    DocumentRangeFormattingProvider = true,
+                    DocumentOnTypeFormattingProvider = new DocumentOnTypeFormattingOptions { FirstTriggerCharacter = "}", MoreTriggerCharacter = new[] { ";", "\n" } },
+                    DefinitionProvider = true,
+                    CompletionProvider = new CompletionOptions { ResolveProvider = true, TriggerCharacters = new[] { "." } },
+                    ImplementationProvider = true,
+                    ReferencesProvider = true,
+                    SignatureHelpProvider = new SignatureHelpOptions { TriggerCharacters = new[] { "(", "," } },
```

Comment:
same issue here.  What about `[`, what about non-c# languages?

Reply:
See above and https://github.com/dotnet/roslyn/issues/40389